File: funcscript/parser/syntax/FuncScriptParser.GetMemberAccess.cs
using funcscript.block;
using funcscript.model;

namespace funcscript.core
{
    public partial class FuncScriptParser
    {
        static ExpressionBlockResult GetMemberAccess(ParseContext context, ExpressionBlock source, int index)
        {
            var result = GetMemberAccessInternal(context, ".", source, index);
            if (result.NextIndex == index)
                return GetMemberAccessInternal(context, "?.", source, index);
            return result;
        }

        static ExpressionBlockResult GetMemberAccessInternal(ParseContext context, string oper, ExpressionBlock source, int index)
        {
            ParseNode parseNode = null;
            ExpressionBlock prog = null;
            var i = SkipSpace(context, index).NextIndex;
            var i2 = GetLiteralMatch(context, i, oper).NextIndex;
            if (i2 == i)
                return new ExpressionBlockResult(null, null, index);

            i = i2;
            i = SkipSpace(context, i).NextIndex;
            var identifierResult = GetIdentifier(context, i, false);

            if (identifierResult.NextIndex == i)
            {
                context.SyntaxErrors.Add(new SyntaxErrorData(i, 0, "member identifier expected"));
                return new ExpressionBlockResult(null, null, index);
            }

            i = identifierResult.NextIndex;
            prog = new FunctionCallExpression
            {
                Function = new LiteralBlock(context.ReferenceProvider.Get(oper)),
                Parameters = new ExpressionBlock[] { source, new LiteralBlock(identifierResult.Iden) },
                CodePos = source.CodePos,
                CodeLength = i - source.CodePos
            };
            prog.SetContext(context.ReferenceProvider);
            return new ExpressionBlockResult(prog, identifierResult.ParseNode, i);
        }
    }
}

File: ts/core/src/parser/syntax/FuncScriptParser.GetMemberAccess.ts
import { ParseContext, ParseNode, ParseNodeType, ParseResult, SyntaxErrorData } from "../FuncScriptParser.Main";
import { GetLiteralMatch } from "./FuncScriptParser.GetLiteralMatch";
import { SkipSpace } from "./FuncScriptParser.SkipSpace";
import { GetIdentifier } from "./FuncScriptParser.GetIdentifier";

export function GetMemberAccess(context: ParseContext, index: number): ParseResult {
    const result = GetMemberAccessInternal(context, ".", index);
    if (result.NextIndex === index) {
        return GetMemberAccessInternal(context, "?.", index);
    }
    return result;
}

function GetMemberAccessInternal(context: ParseContext, oper: string, index: number): ParseResult {
    let parseNode: ParseNode | null = null;

    let i = SkipSpace(context, index).NextIndex;
    let i2 = GetLiteralMatch(context, i, oper).NextIndex;
    if (i2 === i) {
        return { ParseNode: null, NextIndex: index };
    }

    i = i2;
    i = SkipSpace(context, i).NextIndex;

    const identifierResult = GetIdentifier(context, i, false);

    if (identifierResult.NextIndex === i) {
        context.SyntaxErrors.push(
            new SyntaxErrorData(i, 0, "member identifier expected")
        );
        return { ParseNode: null, NextIndex: index };
    }

    i = identifierResult.NextIndex;

    parseNode = new ParseNode(
        ParseNodeType.MemberAccess,
        index,
        i - index,
        [identifierResult.ParseNode!]
    );

    return { ParseNode: parseNode, NextIndex: i };
}